### PR TITLE
fix(upgrade): check latest upgrade

### PR DESCRIPTION
**Pull request by FrankYang0529**
**Problem:**
A previous successful or failed upgrade can interfere with new upgrades.

**Solution:**
Check whether an upgrade has `harvesterhci.io/latestUpgrade: true` label.

**Related Issue:**
https://github.com/harvester/harvester/issues/3152

**Test plan:**

1. Setup a harvester v1.0.3 cluster.
2. Update `rkeConfig` in fleet-local/local `clusters.provisioning.cattle.io`.
```yaml
  rkeConfig:
    provisionGeneration: 0
    upgradeStrategy:
      controlPlaneConcurrency: "1"
      controlPlaneDrainOptions:
        deleteEmptyDirData: true
        enabled: true
        force: true
        ignoreDaemonSets: true
        postDrainHooks:
        - annotation: harvesterhci.io/post-hook
        preDrainHooks:
        - annotation: harvesterhci.io/pre-hook
      workerConcurrency: "1"
      workerDrainOptions:
        deleteEmptyDirData: true
        enabled: true
        force: true
        ignoreDaemonSets: true
        postDrainHooks:
        - annotation: harvesterhci.io/post-hook
        preDrainHooks:
        - annotation: harvesterhci.io/pre-hook
```
3. Run `python3 -m http.server` in your image folder.
4. Create a version. (Remember to change `isoURL`)
```
cat <<EOF | kubectl apply -f -
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: master-head
  namespace: harvester-system
spec:
  isoURL: "http://192.168.0.181:8000/harvester-master-amd64.iso"
EOF
```
5. Create an upgrade.
6. Before VMImage finish, stop the python3 server. Now, we have one failed upgrade in our cluster.
7. Start the python3 server again.
8. Create an upgrade again.
9. Before the upgrade gets into `UpgradingNodes` status, check `rkeConfig` in fleet-local/local `clusters.provisioning.cattle.io` is not changed.

### CODE REVIEW

File: pkg/controller/master/upgrade/upgrade_controller.go [hunk 1]
```diff
@@ -132,6 +133,12 @@ func (h *upgradeHandler) OnChanged(key string, upgrade *harvesterv1.Upgrade) (*h
 		return h.upgradeClient.Update(toUpdate)
 	}
 
+	// only run further operations for latest upgrade
+	if upgrade.Labels == nil || upgrade.Labels[harvesterLatestUpgradeLabel] != "true" {
+		return upgrade, nil
+	}
+
+	logrus.Infof("handle upgrade %s/%s", upgrade.Namespace, upgrade.Name)
```
Comment (starbops) [inline pkg/controller/master/upgrade/upgrade_controller.go]: Is it possible to move this condition checking before L91? If I understand it correctly, we only want to handle the upgrade CR which is the "latest", we can short-circuit others in the early stage.

Comment (FrankYang0529) [reply pkg/controller/master/upgrade/upgrade_controller.go]: We only want to handle latest upgrade CR. However, when users create a new upgrade, there is no `harvesterhci.io/latestUpgrade` on it, so we need L91 to initialize the label. Thanks.

Comment (starbops) [reply pkg/controller/master/upgrade/upgrade_controller.go]: Thanks for the explanation!

**Review by starbops — APPROVED** [review]
Tested on 4-node upgrade, looking good. LGTM, thanks!

File: pkg/controller/master/upgrade/upgrade_controller.go [hunk 2]
```diff
@@ -132,6 +133,12 @@ func (h *upgradeHandler) OnChanged(key string, upgrade *harvesterv1.Upgrade) (*h
 		return h.upgradeClient.Update(toUpdate)
 	}
 
+	// only run further operations for latest upgrade
+	if upgrade.Labels == nil || upgrade.Labels[harvesterLatestUpgradeLabel] != "true" {
```
Comment (guangbochen) [inline pkg/controller/master/upgrade/upgrade_controller.go:139]: prefer to add a log here with proper message and label info.

Comment (FrankYang0529) [reply pkg/controller/master/upgrade/upgrade_controller.go:139]: Added it. Thank you.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**